src/transformations/defmodel.hpp

```diff
@@ -0,0 +1,1890 @@
+                                      double m10, double m11, double &dX,
+                                      double &dY, double &dZ) {
+        // If interpolating in the same cell as before, then we can skip
+        // the recomputation of dXij, dYij and dZij
+        if (ix0 != last_ix0 || iy0 != last_iy0) {
+
+            last_ix0 = ix0;
+            if (iy0 != last_iy0) {
+                const double y0 = grid->miny + iy0 * grid->resy;
+                sinphi0 = sin(y0);
+                cosphi0 = cos(y0);
+
+                // Use trigonometric formulas to avoid new calls to sin/cos
+                sinphi1 =
+                    /* sin(y0+grid->resyRad) = */ sinphi0 * cosresy +
+                    cosphi0 * sinresy;
+                cosphi1 =
+                    /* cos(y0+grid->resyRad) = */ cosphi0 * cosresy -
+                    sinphi0 * sinresy;
+
+                last_iy0 = iy0;
+            }
+
+            const double sinlam00 = -sinhalfresx;
+            const double coslam00 = coshalfresx;
+            const double sinphi00 = sinphi0;
+            const double cosphi00 = cosphi0;
+            const double dn00sinphi00 = dn00 * sinphi00;
+            dX00 = -de00 * sinlam00 - dn00sinphi00 * coslam00;
+            dY00 = de00 * coslam00 - dn00sinphi00 * sinlam00;
+            dZ00 = dn00 * cosphi00;
+
+            const double sinlam01 = -sinhalfresx;
+            const double coslam01 = coshalfresx;
+            const double sinphi01 = sinphi1;
+            const double cosphi01 = cosphi1;
+            const double dn01sinphi01 = dn01 * sinphi01;
+            dX01 = -de01 * sinlam01 - dn01sinphi01 * coslam01;
+            dY01 = de01 * coslam01 - dn01sinphi01 * sinlam01;
+            dZ01 = dn01 * cosphi01;
+
+            const double sinlam10 = sinhalfresx;
+            const double coslam10 = coshalfresx;
+            const double sinphi10 = sinphi0;
+            const double cosphi10 = cosphi0;
+            const double dn10sinphi10 = dn10 * sinphi10;
+            dX10 = -de10 * sinlam10 - dn10sinphi10 * coslam10;
+            dY10 = de10 * coslam10 - dn10sinphi10 * sinlam10;
+            dZ10 = dn10 * cosphi10;
+
+            const double sinlam11 = sinhalfresx;
+            const double coslam11 = coshalfresx;
+            const double sinphi11 = sinphi1;
+            const double cosphi11 = cosphi1;
+            const double dn11sinphi11 = dn11 * sinphi11;
+            dX11 = -de11 * sinlam11 - dn11sinphi11 * coslam11;
+            dY11 = de11 * coslam11 - dn11sinphi11 * sinlam11;
+            dZ11 = dn11 * cosphi11;
+        }
+
+        dX = m00 * dX00 + m01 * dX01 + m10 * dX10 + m11 * dX11;
+        dY = m00 * dY00 + m01 * dY01 + m10 * dY10 + m11 * dY11;
+        dZ = m00 * dZ00 + m01 * dZ01 + m10 * dZ10 + m11 * dZ11;
+    }
+};
+
+// ---------------------------------------------------------------------------
+
+enum class DisplacementType { NONE, HORIZONTAL, VERTICAL, THREE_D };
```

Comment:
I find enums in the middle of long header files difficult to find when scanning a header. Prefer them at the top. (feel free to ignore this)

Reply:
this is mostly an implementation detail (not exposed in public API), so keeping it here